vms/platformvm/validators/fee/fee.go

```diff
@@ -0,0 +1,161 @@
+	current gas.Gas,
+	excess gas.Gas,
+	duration uint64,
+) gas.Gas {
+	if current < target {
+		return excess.SubPerSecond(target-current, duration)
+	}
+	return excess.AddPerSecond(current-target, duration)
+}
+
+// CalculateCost calculates how much to charge based on the dynamic fee
+// mechanism for [duration].
+func CalculateCost(
+	target gas.Gas,
+	current gas.Gas,
+	minPrice gas.Price,
+	excessConversionConstant gas.Gas,
+	excess gas.Gas,
+	duration uint64,
+) uint64 {
+	// If the current and target are the same, the price is constant.
+	if current == target {
+		price := gas.CalculatePrice(minPrice, excess, excessConversionConstant)
+		cost, err := safemath.Mul(duration, uint64(price))
+		if err != nil {
+			return math.MaxUint64
+		}
+		return cost
+	}
+
+	var (
+		cost uint64
+		err  error
+	)
+	for i := uint64(0); i < duration; i++ {
+		excess = CalculateExcess(target, current, excess, 1)
+		// If the excess is 0, the price will remain constant for all future
+		// iterations.
+		if excess == 0 {
+			durationWithZeroExcess := duration - i
+			zeroExcessCost, err := safemath.Mul(uint64(minPrice), durationWithZeroExcess)
+			if err != nil {
+				return math.MaxUint64
+			}
+
+			cost, err = safemath.Add(cost, zeroExcessCost)
+			if err != nil {
+				return math.MaxUint64
+			}
+			return cost
+		}
+
+		price := gas.CalculatePrice(minPrice, excess, excessConversionConstant)
+		cost, err = safemath.Add(cost, uint64(price))
+		if err != nil {
+			return math.MaxUint64
+		}
+	}
+	return cost
+}
+
+// CalculateDuration calculates the duration that it would take to charge at
+// least [targetCost] based on the dynamic fee mechanism. The result is capped
+// at [maxDuration].
+func CalculateDuration(
+	target gas.Gas,
+	current gas.Gas,
+	minPrice gas.Price,
+	excessConversionConstant gas.Gas,
+	excess gas.Gas,
+	maxDuration uint64,
+	targetCost uint64,
+) uint64 {
+	// Because this function can divide by prices, we need to sanity check the
+	// parameters to avoid division by 0.
+	if minPrice == 0 {
+		if targetCost == 0 {
+			return 0
+		}
+		return maxDuration
+	}
+
+	// If the current and target are the same, the price is constant.
+	if current == target {
+		price := gas.CalculatePrice(minPrice, excess, excessConversionConstant)
+		return calculateDuration(
+			uint64(price),
+			targetCost,
+			maxDuration,
+		)
+	}
+
+	var (
+		cost     uint64
+		duration uint64
+		err      error
+	)
+	for cost < targetCost && duration < maxDuration {
+		excess = CalculateExcess(target, current, excess, 1)
+		// If the excess is 0, the price will remain constant for all future
+		// iterations.
+		if excess == 0 {
+			zeroExcessCost := targetCost - cost
+			durationWithZeroExcess := calculateDuration(
+				uint64(minPrice),
+				zeroExcessCost,
+				maxDuration,
+			)
+
+			duration, err = safemath.Add(duration, durationWithZeroExcess)
+			if err != nil || duration >= maxDuration {
+				return maxDuration
+			}
+			return duration
+		}
+
+		duration++
```

Comment:
I think this should be in the loop. As far as I can tell, the only branch it would affect would be the error on `safemath.Add(cost...)`.

```go
for ; cost < targetCost && duration < maxDuration; duration++ {
```

Reply:
Having:
```golang
if err != nil {
	return duration + 1
}
```
feels pretty weird to me... Personally I find the code as-is more readable... but if you have conviction then I'm fine moving the increment around.